lgc/patch/PatchReadFirstLane.cpp

```diff
@@ -0,0 +1,383 @@
+ *  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ *  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ *  SOFTWARE.
+ *
+ **********************************************************************************************************************/
+/**
+ ***********************************************************************************************************************
+ * @file  PatchReadFirstLane.cpp
+ * @brief LLPC source file: contains declaration and implementation of class lgc::PatchReadFirstLane.
+ ***********************************************************************************************************************
+ */
+#include "lgc/Builder.h"
+#include "lgc/patch/Patch.h"
+#include "lgc/state/PipelineState.h"
+#include "llvm/Analysis/LegacyDivergenceAnalysis.h"
+#include "llvm/Analysis/TargetTransformInfo.h"
+#include "llvm/IR/Constants.h"
+#include "llvm/IR/InstVisitor.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/IntrinsicsAMDGPU.h"
+#include "llvm/InitializePasses.h"
+#include "llvm/Pass.h"
+#include "llvm/Support/Debug.h"
+
+#define DEBUG_TYPE "lgc-patch-read-first-lane"
+
+using namespace lgc;
+using namespace llvm;
+
+namespace {
+class PatchReadFirstLane final : public FunctionPass {
+public:
+  PatchReadFirstLane();
+
+  virtual bool runOnFunction(Function &function) override;
+  void getAnalysisUsage(AnalysisUsage &analysisUsage) const override;
+
+  static char ID; // ID of this pass
+
+private:
+  PatchReadFirstLane(const PatchReadFirstLane &) = delete;
+  PatchReadFirstLane &operator=(const PatchReadFirstLane &) = delete;
+
+  bool liftReadFirstLane(Function &function);
+  void collectAssumeUniforms(BasicBlock *block, DenseSet<Instruction *> &initialReadFirstLanes);
+  void findInsertLocations(BasicBlock *block);
+  bool isBestInsertLocation(Instruction *inst);
+  bool isAllUsersAssumedUniform(Instruction *inst, const DenseSet<Instruction *> &assumedUniforms);
+  void applyReadFirstLane(Instruction *inst, BuilderBase &builder);
+  void addLocation(Instruction *inst);
+
+  // We only support to apply amdgcn_readfirstlane on float or int type
+  // TODO: Support various types when backend work is ready
+  bool isSupportedType(Instruction *inst) { return inst->getType()->isFloatTy() || inst->getType()->isIntegerTy(32); }
+
+  LegacyDivergenceAnalysis *m_divergenceAnalysis; // The divergence analysis
+  TargetTransformInfo *m_targetTransformInfo;     // The target tranform info to determine stop propagation
+  DenseMap<Instruction *, DenseSet<Instruction *>>
+      m_canAssumeUniformDivergentUsesMap;    // The instructions that can be assumed uniform
```

Comment:
I suggest making the value type a `SmallVector<Instruction *, 2>`: you don't *really* need it to be a set (even though there could perhaps be duplicates, but not typically), and making it a SmallVector is easier on memory allocations.

Also, can we improve the comment? Perhaps: "The instructions that can be assumed to be uniform. If the vector associated to an instruction I is non-empty, it contains the instructions that, if readfirstlane is applied to them, lead to I being uniform."

Is that accurate? Something like that should make the meaning of the variable clear and helps us reason about it.